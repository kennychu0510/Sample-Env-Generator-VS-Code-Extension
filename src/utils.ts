import * as fs from 'fs';
import * as path from 'path';

export function getFileName(filePath: string) {
  return path.parse(filePath).base;
}

export function generateFile(filePath: string, content: string) {
  fs.writeFile(filePath + '.sample', content, (err) => {
    if (err) {
      throw new Error(err.message);
    }
  });
}

const regex = /^([^=]+)=/;

export function extractContent(filePath: string) {
  const content = fs.readFileSync(filePath, { encoding: 'utf-8' });
  const lines = content.split('\n');
  const outputContent: string[] = lines.map((item) => {
    const match = RegExp(regex).exec(item);
    if (match) {
      return match.at(1) + '=';
    }
    return '';
  });
  return outputContent.join('\n');
}

export function extractContentForZodSchema(filePath: string) {
  const content = fs.readFileSync(filePath, { encoding: 'utf-8' });
  const lines = content.split('\n');
  const outputContent: string[] = lines.map((item) => {
    const match = RegExp(regex).exec(item);
    if (match) {
      const value = item.replace(match.at(1) + '=', '')
      const isNumber = !isNaN(Number(value)) && value.length > 0
      const zodType = isNumber ? 'z.number()' : 'z.string()'
      return '\t' + match.at(1) + `: ${zodType},`
    }
    return '';
  });
  return outputContent.join('\n');
}

export function generateSampleEnv(filePath: string) {
  const content = extractContent(filePath);
  generateFile(filePath, content);
}

export function generateZodSchema(filePath: string) {
  const fileName = path.basename(filePath)
  const content = extractContentForZodSchema(filePath);
  const zodFileContent = zodSchemaStart + content + zodSchemaEnd
  fs.writeFile(filePath.replace(fileName, '') + 'env.ts', zodFileContent, (err) => {
    if (err) {
      throw new Error(err.message);
    }
  });
}

const zodSchemaStart = `import z from 'zod'\n\nconst envSchema = z.object({\n`
const zodSchemaEnd = `\n})\n\nexport const ENV = envSchema.parse(process.env)`