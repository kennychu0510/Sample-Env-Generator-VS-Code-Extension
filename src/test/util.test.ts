import { afterEach, describe, expect, test } from 'vitest';
import * as fs from 'fs'
import { extractContent, extractContentForZodSchema, generateSampleEnv, getFileName } from '../utils';

const env1 = 'src/test/.env'
const emptyEnvFilePath = 'src/test/.env.empty'
const env2 = 'src/test/.env.2'
const env3 = 'src/test/.env.3'
const env4 = 'src/test/.env.4'

describe('get file name', () => {
  test('return file.txt', () => {
    expect(getFileName('/path/to/my/file.txt')).toBe('file.txt')
  });
});

describe('get file content', () => {
  test('extract all the fields correctly', () => {
    expect(extractContent(env1)).toContain(`API_KEY=\nDATABASE_HOST=\nDATABASE_PORT=\nDATABASE_NAME=\nDATABASE_USER=\nDATABASE_PASSWORD=`)
  })
})

describe('empty .env to return empty string', () => {
  test('extract all the fields correctly', () => {
    expect(extractContent(emptyEnvFilePath)).toContain('')
  })
})

describe('env with weird symnbols', () => {
  test('extract all the fields correctly', () => {
    expect(extractContent(emptyEnvFilePath)).toContain('')
  })
})

describe('extract content for zod', () => {
  test('extract type correctly', () => {
    expect(extractContentForZodSchema(env3)).toBeDefined()
  })
})

describe('generate zod file with type correctly', () => {
  test('extract all contents correctly', () => {
    expect(extractContentForZodSchema(env1)).toMatchSnapshot()
    expect(extractContentForZodSchema(env2)).toMatchSnapshot()
    expect(extractContentForZodSchema(emptyEnvFilePath)).toMatchSnapshot()
    expect(extractContentForZodSchema(env4)).toMatchSnapshot()
  })
})


describe('generate env file correctly', () => {
  test('extract all contents correctly', () => {
    expect(extractContent(env1)).toMatchSnapshot()
    expect(extractContent(env2)).toMatchSnapshot()
    expect(extractContent(emptyEnvFilePath)).toMatchSnapshot()
    expect(extractContent(env4)).toMatchSnapshot()
  })
})

describe('generate sample env', () => {
  afterEach(async () => {
    if (fs.existsSync(__dirname + '/.env.sample')) {
      fs.unlinkSync(__dirname + '/.env.sample')
    }
  })
  test('should generate sample env correctly', async () => {
      generateSampleEnv(__dirname + '/.env')
      await sleep(200)
      const output = fs.readFileSync(__dirname + '/.env.sample', {encoding: 'utf-8'})
      expect(output).toMatch(`API_KEY=\nDATABASE_HOST=\nDATABASE_PORT=\nDATABASE_NAME=\nDATABASE_USER=\nDATABASE_PASSWORD=`)
  })
})

function sleep(milliseconds: number) {
  return new Promise((resolve) => {
    setTimeout(resolve, milliseconds);
  });
}