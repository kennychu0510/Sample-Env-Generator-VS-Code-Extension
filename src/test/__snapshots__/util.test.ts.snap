// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`generate env file correctly > extract all contents correctly 1`] = `
"API_KEY=
DATABASE_HOST=
DATABASE_PORT=
DATABASE_NAME=
DATABASE_USER=
DATABASE_PASSWORD="
`;

exports[`generate env file correctly > extract all contents correctly 2`] = `
"API_KEY=
DATABASE_HOST=
DATABASE_PORT=
DATABASE_NAME=
DATABASE_USER=
DATABASE_PASSWORD="
`;

exports[`generate env file correctly > extract all contents correctly 3`] = `""`;

exports[`generate env file correctly > extract all contents correctly 4`] = `
"number=
string=
empty="
`;

exports[`generate zod file with type correctly > extract all contents correctly 1`] = `
"	API_KEY: z.string(),
	DATABASE_HOST: z.string(),
	DATABASE_PORT: z.number(),
	DATABASE_NAME: z.string(),
	DATABASE_USER: z.string(),
	DATABASE_PASSWORD: z.string(),"
`;

exports[`generate zod file with type correctly > extract all contents correctly 2`] = `
"	API_KEY: z.string(),
	DATABASE_HOST: z.string(),
	DATABASE_PORT: z.number(),
	DATABASE_NAME: z.string(),
	DATABASE_USER: z.string(),
	DATABASE_PASSWORD: z.string(),"
`;

exports[`generate zod file with type correctly > extract all contents correctly 3`] = `""`;

exports[`generate zod file with type correctly > extract all contents correctly 4`] = `
"	number: z.number(),
	string: z.string(),
	empty: z.string(),"
`;
